### PR TITLE
Begin work incorporating PoV into Petition pages

### PR DESCRIPTION
**Pull request by dsernst**
This still needs a lot of work, but is the beginning towards displaying a table of individual votes directly on petition pages.

image

### TODO:
- [x] Set width to overflow: scroll so it fits within horiz constraints of page
- [x] Search bar to filter results
- [x] Ensure loading these doesn't block page render 
- [x] Load from cache if available on client side transition
- [x] Load [the other PoVs](https://github.com/voteliquid/proof-of-vote/tree/master/votes), if available (currently hardcoding EDS)
  - [x] Adjust header titles as appropriate (`Senator`/`Assembly` -> `Supervisor` for Dane)
- [ ] Test on mobile (probably bad atm)

### CODE REVIEW

**Comment by dsernst** [comment]
Search bar to filter results now working:

filter backers table

It also stores the current query in the URL `&filter=foobar` to make it easy to link others to specific filters.

**Comment by dsernst** [comment]
Lots of work to speed up loading time.

Next is to load in for other items, instead of only hardcoding EDS.

**Comment by dsernst** [comment]
The core of this is basically now working, for all 3 of the previously published Proofs:

### EducatorsForDemocraticSchools/hold-charter-schools-accountable

Screen Shot 2019-08-22 at 3 11 34 AM

### equalvote/adopt-star-voting

Screen Shot 2019-08-22 at 3 11 38 AM


### prodane/press-pause-on-227m-new-jail

Screen Shot 2019-08-22 at 3 11 41 AM

**Comment by dsernst** [comment]
Loading times could be improved, especially for the EDS one. 

The API only supports fetching 600 results at a time, and requires using `offsets` to traverse the pagination.

In this case it takes 4 network trips, which run in sequence, for a total load time of about 7 seconds.

Some paths forward:
1. Check if the `endorsements_detailed` endpoint can be sped up using an index.
2. Parallelize the individual page queries, instead of running them sequentially.
3. More eagerly show results when they come in. 
    - For example, the first 50 results & total count (content above the fold) could be displayed initially, while the rest of the data loads in the background.

**Comment by dsernst** [comment]
- [ ] Also, switching to API endpoint broke filtering by reps' names (because in a nested field). 
Need to fix.

File: views/endorsement-backers.js [hunk 1]
```diff
@@ -0,0 +1,103 @@
+const { html } = require('../helpers')
+const stateNames = require('datasets-us-states-names-abbr')
+
+const legislatorTitles = {
```
Comment (alexmingoia) [inline views/endorsement-backers.js]: There are more legislators than just California and Dane County... We should use the `name` or `short_name` fields in the database so the naming is consistent and works for every legislator.

File: views/endorsement-backers.js [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        <div>
+          ${searchBar(dispatch, { backers, backersFilterQuery, filteredBackers })}
+          <table class="table is-narrow is-bordered is-striped" style="display: block; overflow-x: auto; max-width: 858px; border-right: 1px solid #e6e6e6">
+            <thead>
+              <tr>
+                <th></th>
+                <th>Time</th>
+                <th>Name</th>
+                <th>Location</th>
+                ${numLegislators > 0 ? html`<th>${titles[0] || 'Legislator'}</th>` : ''}
```
Comment (alexmingoia) [inline views/endorsement-backers.js]: Same here... not all legislatures have two representatives. Many city councils have a bunch of reps. Boise city council there would be six.

File: views/endorsement-backers.js [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+  const name = backer.public ? backer.name : '[private]'
+  const stateAbbr = stateNames[backer.administrative_area_level_1]
+
+  return html`
+    <tr>
+      <td><span style="width: 30px; display: inline-block; text-align: right;">${backer.index}</span></td>
+      <td><span style="width: 165px; display: inline-block;">${new Date(backer.created_at).toLocaleString()}</span></td>
+      <td><span style="width: 165px; display: inline-block;">${name}</span></td>
+      <td><span style="width: 145px; display: inline-block;">${backer.locality}${backer.locality && stateAbbr ? `, ` : ''}${stateAbbr}</span></td>
+      ${numLegislators > 0 ? html`
```
Comment (alexmingoia) [inline views/endorsement-backers.js]: Another instance of hardcoded reps.

File: helpers.js [hunk 4]
```diff
@@ -202,6 +202,9 @@ exports.api = (dispatch, url, params = {}) => {
   if (params.user && params.user.jwt) {
     params.headers.Authorization = `Bearer ${params.user.jwt}`
   }
+  if (params.getAllPages) {
```
Comment (alexmingoia) [inline helpers.js]: We shouldn't be fetching every endorsement... there's thousands of them. Let's just paginate the table 25 at a time.

I optimized the query, but fetching a every endorsement could potentially be megabytes of HTML/JSON.

File: views/endorsement-backers.js (hunk 1 above)
Comment (dsernst) [reply views/endorsement-backers.js]: Agreed. 
This was just a temp fix to match existing Proof docs.
There is a placeholder (generic `Legislator`) in there for places without enumerated titles, but agree would be better to be grabbing this field from the API, ideally in the `offices[]` object.

File: views/endorsement-backers.js (hunk 2 above)
Comment (dsernst) [reply views/endorsement-backers.js]: Yeah. 
Was just trying to meet feature parity with the [existing 3 PoV documents](https://github.com/voteliquid/proof-of-vote/tree/e6d134fa41a839f35ef8159795a7cbb3f0aa889e/votes) for this initial version.

File: helpers.js (hunk 4 above)
Comment (dsernst) [reply helpers.js]: I'm not sure pages of only 25 results each is nearly as useful. Maybe 500?

Part of the appeal is specifically in being able to scroll through each and every entry, seeing just how long it is (and growing).

Comment (dsernst) [reply helpers.js]: Also the filtering logic is all happening on the frontend right now.
So need to move that to backend if only grabbing a partial selection of all the entries.

File: views/endorsement-backers.js (hunk 2 above)
Comment (dsernst) [reply views/endorsement-backers.js]: I updated this to always show the correct number of reps: https://github.com/voteliquid/liquid.us/pull/155/commits/e2ab5510634242ca787ddc3799be788095f33d8e

File: views/endorsement-backers.js (hunk 3 above)
Comment (dsernst) [reply views/endorsement-backers.js]: Fixed by https://github.com/voteliquid/liquid.us/pull/155/commits/e2ab5510634242ca787ddc3799be788095f33d8e

**Comment by dsernst** [comment]
Josh tested:

> Looks good to me

**Pull request merged**